File: Chapter09/Java/ChuckNorrisFacts/app/src/main/java/net/zenconsult/android/chucknorris/Comms.java
package net.zenconsult.android.chucknorris;

import java.io.IOException;
import java.io.InputStream;
import java.util.Random;
import java.util.Vector;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

import org.apache.http.HttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.DefaultHttpClient;
import org.apache.http.util.EntityUtils;
import org.w3c.dom.Document;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import android.util.Log;

public class Comms {

    private static final String TAG = Comms.class.getSimpleName();

    private static final String url = "http://www.chucknorrisfacts.com/";

    private final DefaultHttpClient client;

    public Comms() {
        client = new DefaultHttpClient();
    }

    public String get() {
        InputStream pageStream = doGetAsInputStream(url);
        DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
        DocumentBuilder db;
        Document doc;
        String pageText = "";
        Vector<String> quotes = new Vector<>();
        try {
            db = dbFactory.newDocumentBuilder();
            doc = db.parse(pageStream);
            NodeList nl = doc.getElementsByTagName("div");
            for (int x = 0; x < nl.getLength(); ++x) {
                Node node = nl.item(x);
                NamedNodeMap attributes = node.getAttributes();
                for (int y = 0; y < attributes.getLength(); ++y) {
                    if (attributes.getNamedItem("class") != null) {
                        Node attribute = attributes.getNamedItem("class");
                        if (attribute.getNodeValue()
                                .equals("views-field-title")) {
                            NodeList children = node.getChildNodes();
                            for (int z = 0; z < children.getLength(); ++z) {
                                Node child = children.item(z);
                                if (child.getNodeName()
                                        .equalsIgnoreCase("span"))
                                    quotes.add(child.getTextContent());
                            }
                        }
                    }

                }
            }
            Random r = new Random();
            pageText = quotes.get(r.nextInt(quotes.size() - 1));
            pageStream.close();
        } catch (SAXException | IOException | ParserConfigurationException e) {
            Log.e(TAG, "get=>error: ", e);
        }
        return pageText;
    }

    private String doGetAsString(String url) {
        HttpGet request = new HttpGet(url);
        String result = "";
        try {
            HttpResponse response = client.execute(request);
            int code = response.getStatusLine().getStatusCode();
            if (code == 200) {
                result = EntityUtils.toString(response.getEntity());
            } else {
                Log.e(TAG, "doGetAsString=>Non 200 Status Code " + code);
            }
        } catch (IOException e) {
            Log.e(TAG, "doGetAsString=>error: ", e);
        }
        return result;

    }

    private InputStream doGetAsInputStream(String url) {
        HttpGet request = new HttpGet(url);
        InputStream result = null;
        try {
            HttpResponse response = client.execute(request);
            int code = response.getStatusLine().getStatusCode();
            if (code == 200) {
                result = response.getEntity().getContent();
            } else {
                Log.e(TAG, "doGetAsInputStream=>Non 200 Status Code " + code);
            }
        } catch (IOException e) {
            Log.e(TAG, "doGetAsInputStream=>error: ", e);
        }
        return result;

    }
}

File: Chapter09/Kotlin/ChuckNorrisFacts/app/src/main/java/net/zenconsult/android/chucknorris/Comms.kt
package net.zenconsult.android.chucknorris

import android.util.Log
import org.apache.http.client.methods.HttpGet
import org.apache.http.impl.client.DefaultHttpClient
import org.apache.http.util.EntityUtils
import java.io.InputStream
import java.lang.Exception
import java.util.*
import javax.xml.parsers.DocumentBuilderFactory

class Comms {

    private val client = DefaultHttpClient()

    public fun get(): String {
        var pageText = ""
        val pageStream = doGetAsInputStream(REQUEST_URL)
        pageStream?.let {
            val dbFactory = DocumentBuilderFactory.newInstance()
            val quotes = Vector<String>()
            try {
                val db = dbFactory.newDocumentBuilder()
                val doc = db.parse(it)
                val nl = doc.getElementsByTagName("div")
                for (x in 0 until nl.length) {
                    val node = nl.item(x)
                    for (y in 0 until node.attributes.length) {
                        if (node.attributes.getNamedItem("class") != null) {
                            val attribute = node.attributes.getNamedItem("class")
                            if (attribute.nodeValue.equals("views-field-title")) {
                                for (z in 0 until node.childNodes.length) {
                                    val child = node.childNodes.item(z)
                                    if (child.nodeName.lowercase().equals("span")) {
                                        quotes.add(child.textContent)
                                    }
                                }
                            }
                        }
                    }
                }
                val r = Random()
                pageText = quotes[r.nextInt(quotes.size - 1)]
                pageStream.close()
            } catch (e: Exception) {
                Log.e(TAG, "get=>error: ", e)
            }
        }
        return pageText
    }

    private fun doGetAsString(url: String): String {
        val request = HttpGet(url)
        var result = ""
        try {
            val response = client.execute(request)
            if (response.statusLine.statusCode == 200) {
                result = EntityUtils.toString(response.entity)
            } else {
                Log.e(TAG, "doGetAsString=>Non 200 Status Code ${response.statusLine.statusCode}")
            }
        } catch (e: Exception) {
            Log.e(TAG, "doGetAsString=>error: ", e)
        }
        return result
    }

    private fun doGetAsInputStream(url: String): InputStream? {
        val request = HttpGet(url)
        var result: InputStream? = null
        try {
            val response = client.execute(request)
            if (response.statusLine.statusCode == 200) {
                result = response.entity.content
            } else {
                Log.e(TAG, "doGetAsInputStream=>Non 200 Status Code ${response.statusLine.statusCode}")
            }
        } catch (e: Exception){
            Log.e(TAG, "doGetAsInputStream=>error: ", e)
        }
        return result
    }

    companion object {
        const val TAG = "Comms"
        const val REQUEST_URL = "http://www.chucknorrisfacts.com/"
    }
}